File: fastSlow/fractionalPoint/LLMid.js
import {ListNode} from "../../linkedList/ds.js";

function find_middle(head){
    let slow = head;
    let fast = head;

    // fast.next.next -> for detecting first occurance of middle element in case of even number of nodes
    while(fast && fast.next){
        slow = slow.next;
        fast = fast.next.next;
    }
    return slow;
}

const arr = [1, 2, 3, 4]

let head = new ListNode(arr[0]);
let curr = head;

for(let i=1;i<arr.length;i++){
    curr.next = new ListNode(arr[i]);
    curr = curr.next;
}

let num = find_middle(head);
console.log(num.val);

File: fastSlow/fractionalPoint/cpp/LLMid.cpp
#include<iostream>
#include<vector> 

using namespace std;

struct Node{
    int data;
    Node* next;

    Node(int value){
        data = value;
        next = nullptr;
    }
};

Node* findMidNode(Node* head){
    Node* fast = head;
    Node* slow = head;

    while(fast->next && fast->next->next){
        fast = fast -> next -> next;
        slow = slow -> next;
    }

    return slow;
}

int main(){
    vector<int> nums = {1,2,4,7};

    Node* head = new Node(nums[0]);
    Node* curr = head;

    for(int i=1;i<nums.size();i++){
        curr -> next = new Node(nums[i]);
        curr = curr -> next;
    }

    Node* ans = findMidNode(head);
    cout<<ans->data;
    return 0;
}